### PR TITLE
ROSAENG-61180 | test: add tests for create operatorroles by cluster key

### PR DESCRIPTION
**Pull request by olucasfreitas**
## PR Summary
Add Ginkgo v2 unit tests for `handleOperatorRoleCreationByClusterKey` and helpers in `by_clusterkey.go`, covering 16 test scenarios.

## Detailed Description of the Issue
`cmd/create/operatorroles/by_clusterkey.go` (570L) had **0% coverage**. This is the cluster-key path for `rosa create operator-roles --cluster <name>`, handling STS validation, OIDC provider matching, HCP vs classic branching, and auto/manual mode. This PR adds tests for `validateOperatorRoles` (5 scenarios), `validateOperatorRolesMatchOidcProvider` (3 scenarios), `handleOperatorRoleCreationByClusterKey` (4 scenarios: auto mode, reusable OIDC, HCP path, manual mode), and `computePolicyARN` (4 partition/prefix scenarios).

## Related Issues and PRs
- Jira: [ROSAENG-61180](https://redhat.atlassian.net/browse/ROSAENG-61180)
- Parent: [ROSAENG-60112](https://redhat.atlassian.net/browse/ROSAENG-60112)
- Stack: **3/7** in the [ROSAENG-61180 stack](https://github.com/openshift/rosa/pull/3521)
- Related PR(s): #3521 #3522 #3524 #3525 #3526 #3527

## Type of Change
- [x] test - adds or updates tests only.

## Previous Behavior
`by_clusterkey.go` had no test coverage. Validation and creation logic was only exercised through E2E tests.

## Behavior After This Change
No user-facing behavior change. 16 new unit tests cover the cluster-key operator role creation path including edge cases (no operator roles, OIDC mismatch, HCP cluster, manual mode command generation).

### Preconditions
Go toolchain matching `go.mod`, `make install-hooks` run.

### Test Steps
1. `make test` — full suite passes
2. `go test -v ./cmd/create/operatorroles/...` — all specs pass (16 new + existing)
3. `make lint` — 0 issues

### Expected Results
All tests green, no lint issues.

## Proof of the Fix
- Logs/CLI output: 39 total specs pass in the operatorroles package

## Breaking Changes
- [x] No breaking changes

## Developer Verification Checklist
- [x] Commit subject/title follows `[JIRA-TICKET] | [TYPE]: <MESSAGE>`.
- [x] PR description clearly explains both **what** changed and **why**.
- [x] Relevant Jira/GitHub issues and related PRs are linked.
- [x] `make install-hooks` has been run in this clone.
- [x] Tests were added/updated where appropriate.
- [x] I manually tested the change.
- [x] `make test` passes.
- [x] `make lint` passes.
- [x] `make rosa` passes.
- [x] Documentation or repo-local agent guidance was added/updated where appropriate.
- [x] Any risk, limitation, or follow-up work is documented.

[ROSAENG-61180]: https://redhat.atlassian.net/browse/ROSAENG-61180?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ROSAENG-60112]: https://redhat.atlassian.net/browse/ROSAENG-60112?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive automated coverage for operator-role creation by cluster key.
  * Validated role existence, ARN and policy handling, OIDC paths, automatic and manual creation flows, credential requests, AWS error propagation, and support for varied AWS partitions and paths.

### CODE REVIEW

**Comment by olucasfreitas** [comment]
@coderabbitai full review

**Comment by amandahla** [comment]
@coderabbitai Please perform a focused, read-only review of this PR’s test coverage and behavior.

  Specifically:

  1. Find existing E2E tests related to the changed functionality.
  2. Compare those E2E expectations with the new or modified unit tests. Confirm that the unit tests represent the behavior that is supposed to work in practice.
  3. Verify that mocked OCM API responses accurately reflect the real OCM API models and behavior.
  4. Identify meaningful alternative OCM API results that should be tested, including empty, missing, partial, unexpected, and error responses where applicable.
  5. Map every conditional branch introduced or affected by this PR to a test. Report branches and edge cases that remain uncovered.
  6. Determine whether `buildCommand` can be tested directly or through the constructed Cobra command. Recommend a concrete test structure, including relevant inputs and assertions.
  7. Check failure paths as carefully as success paths, including returned errors and user-visible command behavior.
  8. Do not suggest tests merely to increase coverage. Focus on observable behavior and realistic regressions.

  For every finding, include:

  - Severity and rationale
  - Exact file and line(s)
  - The uncovered scenario or incorrect expectation
  - A concrete recommended test case
  - Whether an existing E2E test already covers the scenario

  Finish with a compact test matrix listing each important condition and whether it is covered by unit tests, E2E tests, both, or neither. If no actionable gaps exist, state that explicitly.

**Comment by amandahla** [comment]
@coderabbitai Please re-evaluate the 4 actionable test gaps identified in your [comment](https://github.com/openshift/rosa/pull/3523#issuecomment-5763115798) above and post each one as an inline review comment on the specific file and line of code where the test should be added or updated. 

For each inline comment, include:
1. The missing test scenario/edge case.
2. A brief suggestion showing how to write the test.

File: cmd/create/operatorroles/by_clusterkey_test.go [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+		BeforeEach(func() {
+			t.RosaRuntime.Creator = &aws.Creator{
+				ARN:       "arn:aws:iam::123456789012:user/test-user",
+				AccountID: "123456789012",
+				Partition: "aws",
+			}
+		})
+
+		Context("auto mode", func() {
+			It("returns nil when all operator roles exist and OIDC config is not reusable", func() {
```
[the remark this answers is not in the document]
Comment (olucasfreitas) [reply cmd/create/operatorroles/by_clusterkey_test.go:311]: Thanks for the suggestion — agree this path is worth covering.

For this PR we're focusing on handler-level coverage for `by_clusterkey.go` without pulling in the full auto-mode `createRoles` pipeline (installer role version, `ForceEnsurePolicy` / `EnsurePolicy`, role attach, etc.). The no-op branch when all roles exist is already tested here, and `ForceEnsurePolicy` has dedicated tests in `pkg/aws/policies_test.go`.

Happy to add force-policy-creation coverage in a follow-up that targets auto-mode role creation end-to-end. Leaving this thread open in case you want to discuss scope or priority.

File: cmd/create/operatorroles/by_clusterkey_test.go [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+				mockAws.EXPECT().CheckRoleExists(gomock.Any()).
+					Return(true, "", nil)
+
+				// GetCredRequests for HCP (is_hypershift=true)
+				t.ApiServer.AppendHandlers(ghttp.CombineHandlers(
+					ghttp.VerifyRequest(http.MethodGet,
+						"/api/clusters_mgmt/v1/aws_inquiries/sts_credential_requests"),
+					ghttp.VerifyFormKV("is_hypershift", "true"),
+					RespondWithJSON(http.StatusOK,
+						`{"kind":"STSCredentialRequestList","page":1,"size":0,"total":0,"items":[]}`),
```
[the remark this answers is not in the document]
Comment (olucasfreitas) [reply cmd/create/operatorroles/by_clusterkey_test.go:534]: Good point that an empty cred-request list plus all roles present exits before HCP-specific policy selection runs.

In this PR the HCP test is scoped to the handler contract we can assert cheaply: `GetCredRequests` is called with `is_hypershift=true` (verified via `VerifyFormKV`). Exercising HCP policy keys through `createRoles` would need the same deep AWS/OCM mocking as a full auto-mode creation test, which we're deferring to keep this change focused.

If you think we should prioritize an HCP auto-mode creation test next, we can track that separately — leaving this open for your take.

File: cmd/create/operatorroles/by_clusterkey_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply cmd/create/operatorroles/by_clusterkey_test.go:311]: The existing tests cover different boundaries:

  - The no-op test covers the default path where all roles exist and forcePolicyCreation is false.
  - pkg/aws/policies_test.go validates ForceEnsurePolicy in isolation.

  This test belongs in by_clusterkey_test.go because it validates the cluster-key command orchestration: with existing roles, unmanaged policies, and forcePolicyCreation=true, the handler must bypass the "Operator Roles already exists" return and dispatch
  to ForceEnsurePolicy rather than EnsurePolicy.

  A test can reuse the existing all-roles-exist fixture, set args.forcePolicyCreation = true, return one credential request, and expect GetAccountRoleVersion, ForceEnsurePolicy, EnsureRole, and AttachRolePolicy. It should not configure an expectation for
  EnsurePolicy; gomock will then fail if that path is selected.

=====

  From the user perspective, this covers:

  rosa create operator-roles \
    --cluster mycluster \
    --mode auto \
    --force-policy-creation

  For an STS cluster with existing operator roles and unmanaged policies, the default command exits with Operator Roles already exists. With --force-policy-creation, it must continue and force the operator policy update.

======

For example, the test can reuse the existing all-roles-exist fixture, set args.forcePolicyCreation = true, return one credential request, and expect specific calls

@coderabbitai is it doable? makes sense? short answer and implementation example.

File: cmd/create/operatorroles/by_clusterkey_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (amandahla) [reply cmd/create/operatorroles/by_clusterkey_test.go:534]: Agree, no problem, closing it as as out of scope.

File: cmd/create/operatorroles/by_clusterkey_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (olucasfreitas) [reply cmd/create/operatorroles/by_clusterkey_test.go:311]: Thanks for pushing on this — agreed it belongs in this PR.

Added `bypasses no-op and calls ForceEnsurePolicy when forcePolicyCreation is true` in the auto-mode context. It uses the existing cluster fixture with all roles present, sets `args.forcePolicyCreation = true`, and mocks `GetAccountRoleVersion`, `ForceEnsurePolicy`, `EnsureRole`, and `AttachRolePolicy`. The strict gomock controller ensures `EnsurePolicy` is not called on this path.

Pushed in the latest commit on this branch.
